pages/base.py

```diff
@@ -31,61 +36,183 @@ def __getattr__(self, element):
+        'user_dropdown_support': (By.CSS_SELECTOR, '#secondary-navigation > ul > li.dropdown.open > ul > li:nth-child(2) > a'),
+        'user_dropdown_settings': (By.CSS_SELECTOR, '#secondary-navigation > ul > li.dropdown.open > ul > li:nth-child(3) > a'),
+        'sign_up_button': (By.LINK_TEXT, 'Sign Up'),
+        'sign_in_button': (By.LINK_TEXT, 'Sign In'),
+        'logout_link': (By.CSS_SELECTOR, '#secondary-navigation > ul > li.dropdown.open > ul > li:nth-child(4) > a'),
+        'current_service': (By.CSS_SELECTOR, '#navbarScope > div > div > div.service-home > a > span.current-service > strong')
+    }
+
+    def verify(self):
+        return len(self.find_elements(By.ID, 'navbarScope')) == 1
+
+    def is_logged_in(self):
+        try:
+            if self.sign_up_button:
+                return False
+        except ValueError:
+            return True
+
+
+class LoginPage(BasePage):
+    url = settings.OSF_HOME + '/login'
+
+    locators = {
+        'identity': (By.XPATH, '/html/body[@id="cas"]/div[@id="container"]', settings.LONG_TIMEOUT),
+        'username_input': (By.ID, 'username'),
+        'password_input': (By.ID, 'password'),
+        'submit_button': (By.NAME, 'submit'),
+        'local_submit_button': (By.ID, 'submit'),
+        'remember_me_checkbox': (By.ID, 'rememberMe'),
+    }
+
+    def __init__(self, driver):
+        super(LoginPage, self).__init__(driver)
+        old_url = driver.current_url
+        driver.get(self.url)
+
+        if not self.verify():
+            url = driver.current_url
+            if url == old_url:
+                raise LoginError(
+                    driver=driver,
+                    error_info='Already logged in'
+                )
+            raise PageException('Unexpected page structure: `{}`'.format(
+                url
+            ))
+
+    def verify(self):
+        try:
+            self.identity
+        except ValueError:
+            return False
+        else:
+            return True
+
+    def login(self, user, password):
+        self.username_input.send_keys(user)
+        if ('localhost:5000' in settings.OSF_HOME):
+            self.local_submit_button.click()
+        else:
+            self.password_input.send_keys(password)
+            if self.remember_me_checkbox.is_selected():
+                self.remember_me_checkbox.click()
+            self.submit_button.click()
+
+
+def login(osf_page, user=settings.USER_ONE, password=settings.USER_ONE_PASSWORD):
+    try:
+        login_page = LoginPage(osf_page.driver)
+    except LoginError:
+        pass
+    else:
+        login_page.login(user, password)
+        osf_page.driver.get(osf_page.url)
+
+
 class OSFBasePage(BasePage):
     url = settings.OSF_HOME
 
-    def __init__(self, driver):
+    # all page must have unique identity
+    locators = {
+        'identity': (By.LINK_TEXT, 'Center for Open Science'),
+        'error_heading': (By.CSS_SELECTOR, 'h2#error'),
+    }
```

Comment:
After talking with @brianjgeiger I and @chennan47  believe that `locators` should be removed from `OSFBasePage`.  

It does not need `identity` because there should never actually be an instance of `OSFBasePage` - opting to use specific pages instead such as `LandingPage` or `DashboardPage`.

`error_heading` can remain in `OSFBasePage` as a property and can call `__getattr__` directly.

This way, when `OSFBasePage` (which it should be almost 100% of the time) is the parent you don't have to do a dictionary merge. And you don't need the `purifyID` util.